runtime/query.go

```diff
@@ -20,6 +21,29 @@ type Query interface {
 }
 
 func (r *Runtime) Query(ctx context.Context, instanceID string, query Query, priority int) error {
-	// TODO: Add caching here
-	return query.Resolve(ctx, r, instanceID, priority)
+	// take all deps and their last updated time and add them to the query key, prefix query key with instanceID
+	cacheKey := instanceID + query.Key()
+	deps := query.Deps()
+	service, err := r.catalogCache.get(ctx, r, instanceID)
+	if err != nil {
+		return err
+	}
+	for _, dep := range deps {
+		entry, found := service.FindEntry(ctx, dep)
+		if !found {
+			r.logger.Error(fmt.Sprintf("dependency %s not found, ignoring it!", dep))
+			continue
+		}
+		cacheKey += entry.Name + entry.UpdatedOn.String()
```

Comment:
Why not just use the instanceID + query key as the cache key, and then *if* there is a cache hit, only then fetch the catalog objects to determine if the cache entry is stale or not?

Reply:
then I will have to store the `updatedOn` as well along with the cached results to compare with latest `updatedOn` to know if the entry is stale or not.

Reply:
It could use wall clock as long as it's single-node, and move to a sequence number in a distributed setting